### PR TITLE
NOTIF-309 Deduplicate incoming ingress Kafka messages

### PR DESCRIPTION
**Pull request by gwenneg**
By default, Kafka guarantees that each message will be consumed/processed _at least once_. This PR introduces Kafka messages deduplication in order to consume each message _exactly once_. The underlying goal is to protect us against Kafka failures like the one we had recently on prod when our consumer failed to commit its offset because it was kicked out of the consumer group. We never confirmed it but it most probably led to duplicate notifications being sent during the incident.

This PR is based on two things:
- A new optional (for now) `rh-message-id` Kafka header which can be sent by the onboarded apps with their message. It is then used as a unique identifier for the message.
- A new `kafka_message` DB table where all valid messages IDs are stored for a limited period of time (default: 1 day, configurable) and then purged.

ℹ️ This is a non-breaking PR that should allow a smooth migration for the onboarded apps. We may decide on a date when the new header will become mandatory, or not. I added a several new counters to allow us to monitor the migration progress. It should help us know when all apps are done migrating.

⚠️ The app behavior should be _exactly identical_ before and after this PR for an onboarded app that _doesn't_ send the new header. If you suspect a behavior change during the review, please add a comment about it.

When a message is received, if its ID is already known (found in the database), then the message is ignored. If the ID is missing (not sent by the onboarded app) or invalid, then the message is considered unknown and processed normally. Invalid IDs are reported at `WARN` level in the logs and will show up in Sentry, allowing us to inform the onboarded app of an integration issue.

⚠️ Version 4 of UUID is mandatory. Any other version received will be rejected.

### CODE REVIEW

File: backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        Optional<KafkaMessageMetadata> metadata = message.getMetadata(KafkaMessageMetadata.class);
+        if (metadata.isPresent()) {
+            Iterator<Header> headers = metadata.get().getHeaders().headers(MESSAGE_ID_HEADER).iterator();
+            if (headers.hasNext()) {
+                found = true;
+                Header header = headers.next();
+                if (header.value() == null) {
+                    invalidMessageIdCounter.increment();
+                    LOGGER.warnf("Application %s/%s sent an invalid Kafka header [%s=null]. They must change their " +
+                                    "integration and send a non-null value.", bundleName, applicationName, MESSAGE_ID_HEADER);
```
Comment (josejulio) [inline backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java:66]: Do we have this header for policies?

Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java:66]: and notifications-gw

Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java:66]: Ah. i see that this is something new.

Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java:66]: Yes it's new.

File: backend/src/main/java/com/redhat/cloud/notifications/models/KafkaMessageId.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+public class KafkaMessageId implements Serializable {
+
+    @NotNull
+    public String bundleName;
+
+    @NotNull
+    public String applicationName;
+
+    @NotNull
+    public UUID messageId;
```
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/models/KafkaMessageId.java:19]: I'm still not sure about the need to store the bundle/app in the database.

If we do that, then we can delegate the responsibility of the `messageId` unicity to the onboarded apps. Duplicates would be their responsibility, not ours.

On the other hand, the chance for a UUID V4 collision is so low that it's very unlikely. Furthermore, the `kafka_message` table is meant to be purged regularly because there's no point storing old messages IDs forever. This reduces even more the collision risk.

Version 4 of UUID is not enforced right now though. They could send us a V3 or lower and it would still be valid for `UUID.fromString`. Maybe we should enforce V4 UUID in the header and get rid of bundle/app in the DB because of the super low collision risk.

WDYT @josejulio?

Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/models/KafkaMessageId.java:19]: cc @pilhuhn

Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/models/KafkaMessageId.java:19]: You will probably need to have a look at `KafkaMessageDeduplicator` to understand the question since the PR description is still missing. 😄

Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/models/KafkaMessageId.java:19]: It would make sense to use only V4... but

This github issue scared me a bit: https://github.com/ramsey/uuid/issues/80
TLDR; There is/was a bug on openssl that increases the chance of collision (if the generator uses a particular function of openssl) when using forked process (collisions across the forked)

Anyhow, the collisions would be for a single application. 
I'm +1 on removing the app/bundle and enforce v4.

File: backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java (hunk 1 above)
Comment (pilhuhn) [reply backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java:66]: I think almost no one is using Kafka headers in all of platform (apps)

File: backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        }
+    }
+
+    public Uni<Void> registerMessageId(String bundleName, String applicationName, UUID messageId) {
+        if (messageId == null) {
+            return Uni.createFrom().voidItem();
+        } else {
+            KafkaMessage kafkaMessage = new KafkaMessage(bundleName, applicationName, messageId);
+            kafkaMessage.prePersist(); // This method must be called manually while using a StatelessSession.
+            return statelessSession.insert(kafkaMessage);
```
Comment (pilhuhn) [inline backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java:126]: Are we cleaning those out?
I think we only need to keep them for a relatively short amount of time (1 day max)

File: backend/src/main/java/com/redhat/cloud/notifications/models/KafkaMessageId.java (hunk 2 above)
Comment (pilhuhn) [reply backend/src/main/java/com/redhat/cloud/notifications/models/KafkaMessageId.java:19]: I wonder if we should store literally as done here, as opposed to use a FK to the apps table, which has a FK to the bundle.

Tbh. what is the collision risk even with that uuid/80 issue? a few hundreds on a million tries? That should be good enough (and the issue is 3-5y old, so most likely fixed)

I think just storing uuid + timestamp could be enough. TS so that we can purge entries older than e.g. a day (I guess the value could be lower)

File: backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java (hunk 3 above)
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java:126]: Yes, that's part of the plan (it's not committed yet though).

File: backend/src/main/java/com/redhat/cloud/notifications/models/KafkaMessageId.java (hunk 2 above)
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/models/KafkaMessageId.java:19]: I don't think we need the relationship between `KafkaMessage` and the app in the database. I added the `bundle`/`application` columns only for unicity purposes but since we all agree that they can be removed, I'll get rid of them. Thanks for your opinion on this @josejulio @pilhuhn!

Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/models/KafkaMessageId.java]: Done.

File: backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java (hunk 3 above)
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java:123]: Cleaner added.

**Comment by gwenneg** [comment]
Description updated, this is ready for a review! 😃

File: backend/src/main/java/com/redhat/cloud/notifications/events/EventConsumer.java [hunk 4]
```diff
@@ -41,56 +46,128 @@
...
                 .onItemOrFailure().transformToUni((unused, throwable) -> {
-                    // If an exception was thrown during steps 1, 2 or 3, an INFO entry with the payload is added to the log.
                     if (throwable != null) {
+                        /*
+                         * An exception was thrown at some point during the Kafka message processing, it is logged.
+                         */
                         LOGGER.infof(throwable, "Could not process the payload: %s", payload);
                     }
+                    /*
+                     * This counter main purpose is to prevent a test race condition.
```
Comment (josejulio) [inline backend/src/main/java/com/redhat/cloud/notifications/events/EventConsumer.java:167]: image

Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/events/EventConsumer.java:167]: ah - nm. It increases on all payloads?  Should the success be counted?

Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/events/EventConsumer.java:167]: They are counter in `EndpointProcessor` 😃

**Review by josejulio — APPROVED** [review]
lgtm
